langchain/logging/base.py

```diff
@@ -0,0 +1,105 @@
+"""Base interface for logging runs."""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict, List, Union
+
+from dataclasses_json import dataclass_json
+
+
+@dataclass_json
+@dataclass
+class Run:
+    id: int
+    start_time: datetime
+    end_time: datetime
+    extra: Dict[str, Any]
+    error: Dict[str, Any]
+    execution_order: int
+    serialized: Dict[str, Any]
+
+
+@dataclass_json
+@dataclass
+class LLMRun(Run):
+    prompts: Dict[str, Any]
+    response: Dict[str, Any]
+
+
+@dataclass_json
+@dataclass
+class ChainRun(Run):
+    inputs: Dict[str, Any]
+    outputs: Dict[str, Any]
+    child_runs: List[Run]
+
+
+@dataclass_json
+@dataclass
+class ToolRun(Run):
+    inputs: Dict[str, Any]
+    outputs: Dict[str, Any]
+    action: str
+    child_runs: List[Run]
+
+
+class BaseLogger(ABC):
+    """Base interface for logging runs."""
+
+    @abstractmethod
+    def log_llm_run_start(
+        self, serialized: Dict[str, Any], prompts: List[str], **extra: str
+    ) -> None:
+        """Log the start of an LLM run."""
+
+    @abstractmethod
+    def log_llm_run_end(self, response: Dict[str, Any], error=None) -> None:
```

Comment:
practically, is error ever used currently

Reply:
not sure response is the best type, should just be LLMResult?

Reply:
error is not used rn, but in the future, it would be good to log the error of a particular run if it didn't succeed.

Agree on LLMResult